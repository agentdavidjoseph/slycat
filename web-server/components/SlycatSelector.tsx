'use strict';
import * as React from 'react';

/**
 * @param {onSelectCallBack} callback when selected
 * @param {options} option list
 * @param {label} label for the dropdown
 */
export interface SlycatSelectorProps { 
  onSelectCallBack: Function
  options: Option[]
  label: string
}

/**
 * object with selector value and text string for display
 *
 * @interface option
 */
export interface Option {
  text: string
  value: string
}

/**
 * not used
 */
export interface SlycatSelectorState {
}
/**
 * class that creates a dropdown with values given in options prop
 */
export default class SlycatSelector extends React.Component<SlycatSelectorProps, SlycatSelectorState> {
  /**
   * not used
   */
  public constructor(props: SlycatSelectorProps) {
    super(props)
    this.state = {};
  }

  /**
   * gets a jsx list of options
   *
   * @memberof SlycatSelector
   */
  getOptions = () => {
    const jsxOptions = this.props.options.map((option, i) => <option key={i} value={option.value}>{option.text}</option>);
    return jsxOptions;
  }

  public render () {
    return (
      <div className="form-group row">
        <label className="col-sm-1 col-form-label">
          {this.props.label}
        </label>
        <div className="col-sm-10">
          <select className="form-control" onChange={(e)=>this.props.onSelectCallBack(e.target.value)}>
            {this.getOptions()}
          </select>
        </div>
      </div>
    )
  }
}
