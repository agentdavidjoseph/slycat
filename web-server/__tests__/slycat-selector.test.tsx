
import React from 'react';
import SlycatSelector, {Option, SlycatSelectorProps, SlycatSelectorState} from "components/SlycatSelector";
import { mount } from "enzyme";

describe('when loading a slycat selector',() =>{

  const options: Option[] = [{
    text:'Comma separated values (CSV)',
    value:'slycat-csv-parser'
  },
  {
    text:'Dakota tabular',
    value:'slycat-dakota-parser'
  }]; 
  const properties: SlycatSelectorProps = {
    onSelectCallBack: jest.fn(),
    label:'Filetype',
    options
  };
  const render = mount(
    <SlycatSelector
      {...properties}
    />
  );
  test('we have expected props on initial load', () => {
    const props = render.props() as SlycatSelectorProps;
    expect(props).toMatchSnapshot();
  });

  test('we see no state', () => {
    const state = render.state() as SlycatSelectorState;
    expect(state).toMatchSnapshot();
  });

  test('we should have a full initialization', () => {
    expect(render.instance()).toMatchSnapshot();
  });

  test('we should onChange function get value', () => {
    const myEventHandler = jest.fn();
    render.setProps({ onSelectCallBack: myEventHandler});
    render.find('select').simulate('change');
    expect(myEventHandler).toHaveBeenCalledWith('slycat-csv-parser');
  });
});

export default undefined