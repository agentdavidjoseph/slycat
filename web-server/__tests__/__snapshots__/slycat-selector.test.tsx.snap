// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when loading a slycat selector we have expected props on initial load 1`] = `
Object {
  "label": "Filetype",
  "onSelectCallBack": [MockFunction],
  "options": Array [
    Object {
      "text": "Comma separated values (CSV)",
      "value": "slycat-csv-parser",
    },
    Object {
      "text": "Dakota tabular",
      "value": "slycat-dakota-parser",
    },
  ],
}
`;

exports[`when loading a slycat selector we see no state 1`] = `Object {}`;

exports[`when loading a slycat selector we should have a full initialization 1`] = `
<SlycatSelector
  label="Filetype"
  onSelectCallBack={[MockFunction]}
  options={
    Array [
      Object {
        "text": "Comma separated values (CSV)",
        "value": "slycat-csv-parser",
      },
      Object {
        "text": "Dakota tabular",
        "value": "slycat-dakota-parser",
      },
    ]
  }
>
  <div
    className="form-group row"
  >
    <label
      className="col-sm-2 col-form-label"
    >
      Filetype
    </label>
    <div
      className="col-sm-9"
    >
      <select
        className="form-control"
        onChange={[Function]}
      >
        <option
          key="0"
          value="slycat-csv-parser"
        >
          Comma separated values (CSV)
        </option>
        <option
          key="1"
          value="slycat-dakota-parser"
        >
          Dakota tabular
        </option>
      </select>
    </div>
  </div>
</SlycatSelector>
`;
